handlejoin.go

```diff
@@ -0,0 +1,278 @@
+package gomatrixserverlib
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/matrix-org/util"
+)
+
+type HandleMakeJoinInput struct {
+	Context            context.Context
+	UserID             spec.UserID
+	RoomID             spec.RoomID
+	RoomVersion        RoomVersion
+	RemoteVersions     []RoomVersion
```

Comment:
Unclear why there are 2 sets of versions here. I guess the `[]` is "I know these versions" but why is it dictating what version the room is to join? How does it even know that info pre make_join?

Reply:
Because this is the receiving side